shard/committee/assignment.go

```diff
@@ -0,0 +1,261 @@
+				pubKey,
+				nil,
+			}
+			com.NodeList = append(com.NodeList, curNodeID)
+		}
+		shardState = append(shardState, com)
+	}
+	return shardState
+}
+
+func with400Stakers(
+	s shardingconfig.Instance, stakerReader StakingCandidatesReader,
+) (shard.State, error) {
+	// TODO Nervous about this because overtime the list will become quite large
+	candidates := stakerReader.ValidatorCandidates()
+	stakers := make([]*staking.Validator, len(candidates))
+	for i := range candidates {
+		// TODO Should be using .ValidatorStakingWithDelegation, not implemented yet
+		validator, err := stakerReader.ValidatorInformation(candidates[i])
+		if err != nil {
+			return nil, err
+		}
+		stakers[i] = validator
+	}
+
+	sort.SliceStable(
+		stakers,
+		func(i, j int) bool { return stakers[i].Stake.Cmp(stakers[j].Stake) >= 0 },
+	)
+	const sCount = 401
+	top := stakers[:sCount]
+	shardCount := int(s.NumShards())
+	superComm := make(shard.State, shardCount)
+	fillCount := make([]int, shardCount)
+	// TODO Finish this logic, not correct, need to operate EPoS on slot level,
+	// not validator level
+
+	for i := 0; i < shardCount; i++ {
+		superComm[i] = shard.Committee{}
+		superComm[i].NodeList = make(shard.NodeIDList, s.NumNodesPerShard())
+	}
+
+	scratchPad := &bls.PublicKey{}
+
+	for i := range top {
+		spot := int(top[i].Address.Big().Int64()) % shardCount
+		fillCount[spot]++
+		// scratchPad.DeserializeHexStr()
+		pubKey := shard.BlsPublicKey{}
+		pubKey.FromLibBLSPublicKey(scratchPad)
+		superComm[spot].NodeList = append(
+			superComm[spot].NodeList,
+			shard.NodeID{
+				top[i].Address,
+				pubKey,
+				&shard.StakedMember{big.NewInt(0)},
+			},
+		)
+	}
+
+	utils.Logger().Info().Ints("distribution of Stakers in Shards", fillCount)
+	return superComm, nil
+}
+
+func (def partialStakingEnabled) ReadPublicKeysFromChain(
+	h common.Hash, reader ChainReader,
+) ([]*bls.PublicKey, error) {
+	header := reader.GetHeaderByHash(h)
+	shardID := header.ShardID()
+	superCommittee, err := reader.ReadShardState(header.Epoch())
+	if err != nil {
+		return nil, err
+	}
+	subCommittee := superCommittee.FindCommitteeByID(shardID)
+	if subCommittee == nil {
+		return nil, ctxerror.New("cannot find shard in the shard state",
+			"blockNumber", header.Number(),
+			"shardID", header.ShardID(),
+		)
+	}
+	committerKeys := []*bls.PublicKey{}
+
+	for i := range subCommittee.NodeList {
+		committerKey := new(bls.PublicKey)
+		err := subCommittee.NodeList[i].BlsPublicKey.ToLibBLSPublicKey(committerKey)
+		if err != nil {
+			return nil, ctxerror.New("cannot convert BLS public key",
+				"blsPublicKey", subCommittee.NodeList[i].BlsPublicKey).WithCause(err)
+		}
+		committerKeys = append(committerKeys, committerKey)
+	}
+	return committerKeys, nil
+
+	return nil, nil
+}
+
+// ReadPublicKeysFromChain produces publicKeys of entire supercommittee per epoch, optionally providing a
+// shard specific subcommittee
+func (def partialStakingEnabled) ReadPublicKeysFromComputation(
```

Comment:
ComputePublicKeys. Read means directly fetch data without doing anything. Read...Computation is confusing. No need to follow the same pattern with the above method.